providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,631 @@
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int ml_kem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
+    vinfo_t v = key->vinfo;
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (params == NULL)
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
+        return 1;
+
+    /* Key mutation is reported generally not allowed */
+    if (have_keys(key)) {
+        ERR_raise_data(ERR_LIB_PROV,
+                       PROV_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE,
+                       "ML-KEM keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    if (publen != v->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+    if (ossl_ml_kem_vparse_public_key(v, &key->pubkey, pubenc, publen,
+                                      key->ctx) == 1)
+        return 1;
+
+    OPENSSL_free(key->pubkey);
+    key->pubkey = NULL;
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+    return 0;
+}
+
+static int ml_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    struct ml_kem_gen_ctx *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        gctx->propq = OPENSSL_strdup(p->data);
+        if (gctx->propq == NULL)
+            return 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ML_KEM_SEED);
+    if (p == NULL)
+        return 1;
+
+    /* Treat wrong data type as promised, but missing */
+    if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SEED);
+        return 0;
+    }
+    if (p->data_size != ML_KEM_SEED_BYTES) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_SEED_LENGTH);
+        return 0;
+    }
+
+    gctx->seed = gctx->seedbuf;
+    memcpy(gctx->seed, p->data, ML_KEM_SEED_BYTES);
+
+    return 1;
+}
+
+static void *ml_kem_gen_init(void *provctx, vinfo_t v,
+                             int selection, const OSSL_PARAM params[])
+{
+    struct ml_kem_gen_ctx *gctx = NULL;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((gctx = OPENSSL_zalloc(sizeof(*gctx))) == NULL)
+        return NULL;
+
+    gctx->provctx = provctx;
+    gctx->vinfo = v;
+    gctx->selection = selection;
+
+    if (!ml_kem_gen_set_params(gctx, params)) {
+        OPENSSL_free(gctx->propq);
```

Comment:
Seems like a wierd place to free propq here.. It should have been handled in set_params()

Reply:
It is simpler to handle deallocation of the context on error in `gen_init()`. This is done by calling `ml_kem_gen_cleanup()` in #26172